### PR TITLE
Add new CLI command 'list-nodes' to list and filter Jenkins nodes

### PR DESCRIPTION
**Pull request by ahmedanwar1**
See [JENKINS-75770](https://issues.jenkins.io/browse/JENKINS-75770).

This pull request introduces a new CLI command called list-nodes into the Jenkins core.

The goal is to provide a more powerful and script-friendly way to retrieve Jenkins node information from the command line, supporting use cases like infrastructure monitoring, automation, and troubleshooting. The command works with both the controller and agent nodes and includes fine-grained filtering options and permission-sensitive output formats.

The CLI command supports the following options:

Filtering options:
- `-status <ONLINE|OFFLINE|ALL>`: Filter nodes by their online/offline status.

- `-labels <label1,label2,...>`: Filter by comma-separated labels (case-insensitive).

- `-labels-mode <ANY|ALL>`: Choose whether any or all specified labels must match.

- `-min-executors <N>`: Only include nodes with at least N executors.

- `-exclude-controller`: Exclude the built-in controller (master) node.

- `-limit <N>`: Return only up to N nodes.

Output options:
- `-format <PLAIN|JSON>`: Choose between human-readable (default) and JSON output.

- `-name-only`: Show only node names (one per line).

- `-count-only`: Show only the number of matching nodes.

- `-verbose`: Display additional system and runtime details per node, such as: OS, architecture, Java version, Offline cause, Clock difference, Connection timestamp. (Requires `Computer.EXTENDED_READ` permission per node).

The output adapts based on user permissions. If `EXTENDED_READ` is missing for a node, verbose fields are replaced with "Permission Denied" for that node.

This contribution uses the existing Jenkins CLI infrastructure and follows security and formatting guidelines.

### Testing done

A dedicated test class `ListNodesCommandTest` has been implemented. It provides comprehensive automated test coverage using `JenkinsRule` and `CLICommandInvoker`.

The tests include:
-   ✅ Basic listing functionality
-   ✅ Filtering by:
    -   Status (online/offline/all)
    -   Labels and label modes (ANY/ALL)
    -   Minimum executors
    -   Excluding the controller
    -   Result limits
-   ✅ Output mode tests:
    -   JSON and plain text
    -   Name-only
    -   Count-only   
-   ✅ Verbose output for both online and offline nodes
-   ✅ Permission-sensitive tests:
    -   Behavior with/without `Jenkins.READ`
    -   Behavior with/without `Computer.EXTENDED_READ` on different nodes
-   ✅ JSON structure validation and edge case handling (e.g. no matches)
-   ✅ Combination tests using multiple filters together
 
All key features and edge cases are covered and validated under automated tests.

### Proposed changelog entries

-   Add new `list-nodes` CLI command to list and filter Jenkins nodes
-   Support output customization (plain, JSON, name-only, count-only)
-   Add filtering options for node status, labels, executors, and more
-   Introduce verbose mode to display system and runtime details for each node (permission-aware)

### Proposed changelog category

/label developer

### Proposed upgrade guidelines

N/A

### Submitter checklist

- [X] The Jira issue, if it exists, is well-described.
- [X] The changelog entries and upgrade guidelines are appropriate for the audience affected by the change (users or developers, depending on the change) and are in the imperative mood (see [examples](https://github.com/jenkins-infra/jenkins.io/blob/master/content/_data/changelogs/weekly.yml)). Fill in the **Proposed upgrade guidelines** section only if there are breaking changes or changes that may require extra steps from users during upgrade.
- [X] There is automated testing or an explanation as to why this change has no tests.
- [X] New public classes, fields, and methods are annotated with `@Restricted` or have `@since TODO` Javadocs, as appropriate.
- [ ] New deprecations are annotated with `@Deprecated(since = "TODO")` or `@Deprecated(forRemoval = true, since = "TODO")`, if applicable.
- [ ] New or substantially changed JavaScript is not defined inline and does not call `eval` to ease future introduction of Content Security Policy (CSP) directives (see [documentation](https://www.jenkins.io/doc/developer/security/csp/)).
- [ ] For dependency updates, there are links to external changelogs and, if possible, full differentials.
- [ ] For new APIs and extension points, there is a link to at least one consumer.

### Desired reviewers

@jenkinsci/core-pr-reviewers

Before the changes are marked as `ready-for-merge`:

### Maintainer checklist

- [ ] There are at least two (2) approvals for the pull request and no outstanding requests for change.
- [ ] Conversations in the pull request are over, or it is explicit that a reviewer is not blocking the change.
- [ ] Changelog entries in the pull request title and/or **Proposed changelog entries** are accurate, human-readable, and in the imperative mood.
- [ ] Proper changelog labels are set so that the changelog can be generated automatically.
- [ ] If the change needs additional upgrade steps from users, the `upgrade-guide-needed` label is set and there is a **Proposed upgrade guidelines** section in the pull request title (see [example](https://github.com/jenkinsci/jenkins/pull/4387)).
- [ ] If it would make sense to backport the change to LTS, a Jira issue must exist, be a _Bug_ or _Improvement_, and be labeled as `lts-candidate` to be considered (see [query](https://issues.jenkins.io/issues/?filter=12146)).

### CODE REVIEW

**Comment by mawinter69** [comment]
I think that there should be the possibility to ask for agents that are taken offline temporarily.
If you want to be absolutely correct an agent can have much more states.
Online and accepting tasks
Online and not accepting task (e.g. an inbound agent that uses a time based availability and is currently outside this window)
Connected but temporarily disabled
disconnected because the availability disallows it (e.g. an outbound agent that uses a time based availability and is currently outside this window)
disconnected because of some problem e.g. network issue, machine crash or agent process crashed

Probably doesn't make sense to have that fine grained distinction in the cli, but temporarily offline I think is a must including potentially the reason why it is offline (thought the reason might contain html I think)

**Comment by ahmedanwar1** [comment]
@mawinter69 Thanks a lot for the detailed feedback it's really helpful

I would keep the existing `OFFLINE` option to cover all offline nodes and introduce an additional `TEMP_OFFLINE` status to specifically target agents that are temporarily taken offline

would that be a good starting point or should I consider going further and support distinguishing additional states like “online but not accepting tasks” as well?

Also thank you for pointing out that offline causes might contain html. would it be best to strip them entirely and converted to plain text (removing tags but preserving the content) or escaping the html, or just left as is in html format

**Comment by mawinter69** [comment]
> I would keep the existing `OFFLINE` option to cover all offline nodes and introduce an additional `TEMP_OFFLINE` status to specifically target agents that are temporarily taken offline
> 
> would that be a good starting point or should I consider going further and support distinguishing additional states like “online but not accepting tasks” as well?

That's a good starting point. That fine grained distinction is probably too confusing as most users are probably not aware of those states. 

About the html probably leave as is, not too many people will use html when taking an agent offline and it also depends on the markup formatter if it is displayed properly on the agent page. And it could also be markdown theoretically if someone installed the markdown formatter plugin and configured it globally.

**Comment by ahmedanwar1** [comment]
Thanks a lot @mawinter69 . I've added the `TEMP_OFFLINE` option to list temporarily offline nodes and adjusted the existing `OFFLINE` option to include all offline nodes (including temporary ones). I also updated the test class to cover these changes and included the offline cause reason in the standard CLI output.

File: core/src/main/java/jenkins/cli/ListNodesCommand.java [hunk 1]
```diff
@@ -0,0 +1,578 @@
...
+    /**
+     * Filter nodes by status: ONLINE, OFFLINE (all offline nodes, including temporarily offline), TEMP_OFFLINE (temporarily offline nodes only), or ALL (default).
+     */
+    @Option(name = "-status", usage = "Filter nodes by status: ONLINE, OFFLINE (all offline nodes, including temporarily offline), TEMP_OFFLINE (temporarily offline nodes only), or ALL")
+    private Status status = Status.ALL;
+
+    /**
+     * Filter nodes by labels, using a comma-separated list (no spaces) (e.g., label1,label2,label3).
+     */
+    @Option(name = "-labels", usage = "Filter nodes by labels (comma-separated, no spaces) (e.g., label1,label2,label3)", metaVar = "label1,label2,...")
```
Comment (mawinter69) [inline core/src/main/java/jenkins/cli/ListNodesCommand.java]: I think it would be better if we can use label expressions. That is what most people know as you already use it when you assign a job or a pipeline to an agent. This us much more powerful and makes the `-labels-mode` option obsolete

File: core/src/main/java/jenkins/cli/ListNodesCommand.java [hunk 2]
```diff
@@ -0,0 +1,578 @@
...
+     * @param millis The timestamp in milliseconds.
+     * @return A formatted date-time string or "N/A" if the timestamp is invalid.
+     */
+    private String formatTimestamp(long millis) {
+        if (millis <= 0) {
+            return "N/A";
+        }
+
+        Instant instant = Instant.ofEpochMilli(millis);
+        DateTimeFormatter formatter = DateTimeFormatter.ofPattern("yyyy-MM-dd HH:mm:ss")
```
Comment (mawinter69) [inline core/src/main/java/jenkins/cli/ListNodesCommand.java]: I would include in the formatted string information about the zone offset by adding `Z`. It avoids confusion when the controller is not running in UTC timezone.

**Review by mawinter69 — CHANGES_REQUESTED** [review]

File: core/src/main/java/jenkins/cli/ListNodesCommand.java (hunk 1 above)
Comment (ahmedanwar1) [reply core/src/main/java/jenkins/cli/ListNodesCommand.java]: great idea that would be better than the current approach

File: core/src/main/java/jenkins/cli/ListNodesCommand.java (hunk 2 above)
Comment (ahmedanwar1) [reply core/src/main/java/jenkins/cli/ListNodesCommand.java]: sure i will add the zone offset in the formatted string

**Comment by ahmedanwar1** [comment]
I replaced comma-separated label filtering with label expressions, and added timezone info (Z) to timestamp formatting

**Comment by ahmedanwar1** [comment]
Hi, it's been almost a month without feedback. just checking in to see if this feature aligns with the current direction of the project, or if I should consider closing the PR.

**Review by mawinter69 — APPROVED** [review]

**Review by A1exKH — APPROVED** [review]
LGTM.

File: core/src/main/java/jenkins/cli/ListNodesCommand.java [hunk 3]
```diff
@@ -0,0 +1,572 @@
...
+     * @return true if the computer matches the status filter, false otherwise.
+     */
+    private boolean matchesStatus(Computer computer) {
+        switch (status) {
+            case TEMP_OFFLINE:
+                return computer.isTemporarilyOffline();
+            case ONLINE:
+                return computer.isOnline();
+            case OFFLINE:
+                return !computer.isOnline();
```
Comment (rsandell) [inline core/src/main/java/jenkins/cli/ListNodesCommand.java]: negation of a negation https://github.com/jenkinsci/jenkins/blob/c5b876b582ab2e5676670952cbc089ef71aa368d/core/src/main/java/hudson/model/Computer.java#L625
technically nothing wrong with it as I believe the compiler fixes it, but
```suggestion
                return computer.isOffline();
```
reads better.

Comment (ahmedanwar1) [reply core/src/main/java/jenkins/cli/ListNodesCommand.java]: Good point, thanks. I'll update it to use `computer.isOffline()` instead.

**Comment by ahmedanwar1** [comment]
Hi @MarkEWaite , I hope you're doing well! I wanted to ask if you have any thoughts on the current state of this PR when you have a chance. I'd really appreciate any feedback. Thanks for your time
